src/azure-cli-core/tests/test_progress.py

```diff
@@ -0,0 +1,153 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+# --------------------------------------------------------------------------------------------
+import unittest
+import mock
+
+import azure.cli.core.commands.progress as progress
+
+
+class MockOutstream(progress._ProgressViewBase):
+    """ mock outstream for testing """
+    def __init__(self, p_type=progress.ProgressType.Indeterminate):
+        self.string = ''
+        self.type = p_type
+
+    def write(self, message):
+        self.string = message
+
+    def get_type(self):
+        return self.type
+
+    def flush(self):
+        return self.string
+
+
+class DetMockOutstream(progress._ProgressViewBase):
+    """ mock outstream for testing """
+    def __init__(self, p_type=progress.ProgressType.Determinate):
+        self.string = ''
+        self.type = p_type
+
+    def write(self, message):
+        self.string = message
+
+    def get_type(self):
+        return self.type
+
+    def flush(self):
+        return self.string
+
+
+
+class TestProgress(unittest.TestCase):  # pylint: disable=too-many-public-methods
+    """ test the progress reporting """
+
+    def test_det_model(self):
+        """ test the determinate progress reporter """
+        reporter = progress.DetProgressReporter()
+        message, percent = reporter.report()
+        self.assertEqual(message, '')
+        self.assertEqual(percent, None)
+
+        reporter.add('Progress', 0, 10)
+        self.assertEqual(reporter.message, 'Progress')
+        self.assertEqual(reporter.curr_val, 0)
+        self.assertEqual(reporter.total_val, 10)
+        message, percent = reporter.report()
+        self.assertEqual(message, 'Progress')
+        self.assertEqual(percent, 0.0)
+
+        with self.assertRaises(AssertionError):
+            reporter.add('In words', -1, 10)
+        with self.assertRaises(AssertionError):
+            reporter.add('In words', 1, -10)
+        with self.assertRaises(AssertionError):
+            reporter.add('In words', 490, 10)
+
+    def test_indet_model(self):
+        """ test the indetermiante progress reporter """
+        reporter = progress.InDetProgressReporter()
+        message = reporter.report()
+        self.assertEqual(message, '')
+
+        reporter.add('Progress')
+        self.assertEqual(reporter.message, 'Progress')
+
+        message = reporter.report()
+        self.assertEqual(message, 'Progress')
+
+    def test_indet_stdview(self):
+        """ tests the indeterminate progress standardout view """
+        view = progress.IndeterminateStandardOut()
+        self.assertEqual(view.get_type, progress.ProgressType.Indeterminate)
+        before = view.spinner.total
+        self.assertEqual(view.spinner.label, 'In Progress')
+        view.write({})
+        after = view.spinner.total
+        self.assertTrue(after > before)
+        view.write({'message':'TESTING'})
```

Comment:
This is why I'd suggest keyword arguments.
`view.write(message='TESTING')` is a lot clearer than what you have.